tasks/broad/Dragen.wdl

```diff
@@ -0,0 +1,102 @@
+version 1.0
+
+## Copyright Broad Institute, 2020
+##
+## This WDL defines tasks to use Dragen's DRAGstr approach to STR sequencing artifacts 
+## Indel genotype priors in the DRAGEN-Gatk pipeline. 
+##
+## Runtime parameters are often optimized for Broad's Google Cloud Platform implementation.
+## For program versions, see docker containers.
+##
+## LICENSING :
+## This script is released under the WDL source code license (BSD-3) (see LICENSE in
+## https://github.com/broadinstitute/wdl). Note however that the programs it calls may
+## be subject to different licenses. Users are responsible for checking that they are
+## authorized to run all programs before running this script. Please see the docker
+## page at https://hub.docker.com/r/broadinstitute/genomes-in-the-cloud/ for detailed
+## licensing information pertaining to the included programs.
+
+
+task ComposeSTRTableFile {
+    
+  input {
+    File ref_fasta
+    File ref_fasta_idx
+    File ref_dict
+    String docker = "us.gcr.io/broad-dsde-methods/broad-gatk-snapshots/dragen_final_test_v2"
+    Int preemptible_tries = 3
+  }
+
+  String base_name = sub(basename(ref_fasta), ".fa(sta)$","")
+  String str_table_file_name = base_name + ".str"
+  Int disk_size_gb = ceil(size([ref_fasta, ref_fasta_idx], "GiB")) + 
+                        21 # 1gb for the output file (usually far less than that) and 20 for the rest of the fs.
+
+  command <<<
+
+    gatk --java-options "-Xmx2g" \
+      ComposeSTRTableFile \
+        -R ~{ref_fasta} \
+        -O ~{str_table_file_name}
+
+  >>>
+
+  runtime {
+    docker: docker 
+    disks: "local-disk " + disk_size_gb + " HDD"
+    memory: "3 GiB"
+    preemptible: preemptible_tries
+  }
+
+  output {
+    File str_table_file = "~{str_table_file_name}"
+  }
+}
+
+task CalibrateDragstrModel {
+
+  input {
+    File ref_fasta
+    File ref_fasta_idx
+    File ref_dict
+    File str_table_file
+    File alignment ## can handle cram or bam.
+    File alignment_index
+    String docker = "us.gcr.io/broad-gatk/gatk:4.2.0.0"
+    Int preemptible_tries = 3
+    Int threads = 4
+    Boolean use_ssd = true
+  }
+
+  String base_name = basename(alignment)
+  String out_file_name = base_name + ".dragstr"
+  Int disk_size_gb = ceil(size([ref_fasta, ref_fasta_idx, alignment, str_table_file], "GiB")) + 
+                        20 # 20 for the rest of the fs.
+
+  String parallel_args  = if (threads <= 1) then "" else "--parallel" 
+
+  command <<<
+    set -x
+    gatk --java-options "-Xmx2g -DGATK_STACKTRACE_ON_USER_EXCEPTION=true -Dsamjdk.reference_fasta=~{ref_fasta}" \
```

Comment:
This isn't enough in a lot of cases.  Mem should be exposed as a param.

Reply:
Right, it looks like (almost) none of the JVM memory limits are exposed in WARP, but I can certainly do that here. Thanks for noticing. From your experience, do you have a point of reference on what would work as a default value?

Reply:
Doing some single sample testing its seems that parallel does not work well in the input is a cram file unless the memory is push up, perhaps as much as 2gb for each intended parallel thread. And even so it might be nearly as fast to transform the cram into bam (using. samtools view -@ thread-count...). bam with or without parallelism does not seem to be affected and in that particular case they can run with the default 2gb without issues.

Reply:
Thanks for your input. It seems like in this pipeline this task is only ever called with BAM input, so I wonder why Laura would run into issues. For me, this task has also never failed, but if tasks fail we should absolutely fix it. @ldgauthier could you maybe confirm if this task failed for you within this pipeline or when executed separately (e.g. on a CRAM)?

Reply:
if accepts cram we should make it does work with them, that reasonable assumption is what got @ldgauthier in trouble.

Reply:
I have posted in my fork changes for the calibrate task to fix the issue with cram: https://github.com/vruano/warp/pull/1

This is based on some testing I performed on a. problematic sample data HG00544

Reply:
Great, thanks @vruano! I added the changes to this branch. I also gave you access to this repo, feel free to merge or PR-merge into this branch as well

Reply:
It turns out the OOM issue is cram-related: https://github.com/broadinstitute/gatk/issues/7189#issuecomment-817806008
My suggestion is we restrict input to bams (just in the docs, nothing fancy)

Reply:
But do still bubble up the mem argument

Reply:
What's the latest on this thread?

Reply:
The memory argument has been exposed, so if I understand correctly then we can resolve this thread? Or are there concerns about this task as it stands right now @ldgauthier ? For reference, this task now lives in https://github.com/michaelgatzen/warp/blob/3e41edc997460dfd42dc53b4cbef4731c0751eda/tasks/broad/DragenTasks.wdl#L19